### DISCUSSION TITLE
Dynamic rules() initialization error

### DISCUSSION THREAD

**Discussion by ju5t** [livewire/livewire#2258] [1. Help] [answered]
I'm trying to make dynamic validation work in a Livewire component. I know you can use `rules()` for that, but first let me walk you to through our component. The route looks like this:

```php
Route::get('quotes/{quote}', QuoteEdit::class)->name('admin.quote.edit');
```

`{quote}` will be a number. In `QuoteEdit` we have a public property for `$quote` that Livewire initiates:

```php
public Quote $quote;
```

This all works well. But these quotes are linked to a dynamic sets of fields. Let's say a quote is made for a new type of paint. We would store some characteristics of that paint in that quote. If it is a quote for a new type of chair, we store some details of this chair with the quote. 

We do this with polymorphic relations. The attribute on the model for this relationship is called `detailable_by` and contains the class name of the `Model` that holds all the attributes for a specific product type (for example: `App\Models\Paint`).

To make our validation rules dynamic, we have a `rules()` method.

Part of this method is the following code:

```php
collect($this->quote->detailable_by::rules())->mapWithKeys(function ($key, $item) {
    return ['detail.'.$item => $key];
});
```

We do this, because the actual key/values of the relation (for `App\Models\Paint`) are stored in the `detail` property of the Livewire component, so we have to prefix the validation rules `detail` for them to validate properly. 

In theory, this should work.

But it doesn't. Whenever I try to run the validation with `$this->validate();` it returns:

```
$quote must not be accessed before initialization
```

This confuses me. We use `$this->quote` as it should be hydrated by Livewire. I've got `quote.detailably_by` in our `rules()` and there are no errors when loading the component initially. It only happens when `$this->validate()` gets triggered.

I also tried to set another public property on the component and give it the `detailably_by` class (a string that is). This gives the same error. What am I missing here?

As this is an important piece for this app to work, if you know a solution, let me know if I can get you a cup of coffee or sponsor something else that can help you.

**Comment by joshhanley** [comment, marked as answer]
@ju5t yeah that is a common issue, Caleb has PR'd a fix #2248 which should be available in the next release 🙂

What you can do, as a work around though, is remove any of the dynamic rules out of your rules method, and instead just pass them into a second validation method on your save/update method. What will then happen is it will run through both set of rules ensuring its all valid.

```php
public function save()
{
    $this->validate(); //Uses any rules in the validation array/method
    $this->validate([
        // Put any rules that need access to dynamic data here
    ]);

    // Do whatever else you then need to here.
}
```

Hope this helps!

**Reply by ju5t** [reply]
It definitely does, thanks a lot @joshhanley!

**Reply by joshhanley** [reply]
No worries! 🙂
